File: demos/lua/dsa/dp/knapsack.lua
local function knapsack(W, weights, values)
    local dp = {}
    for i = 0, W do dp[i] = 0 end

    for i = 1, #weights do
        local w = weights[i]
        local v = values[i]
        for j = W, w, -1 do
            local candidate = dp[j - w] + v
            if candidate > dp[j] then
                dp[j] = candidate
            end
        end
    end
    return dp[W]
end

print("=== Lua 0/1 Knapsack DP ===")
local weights = {2, 3, 4, 5}
local values = {3, 4, 5, 6}
local max_val = knapsack(5, weights, values)
assert(max_val == 7, "Knapsack assertion failed")

print("Max Knapsack Value for W=5: " .. max_val)
print("Lua Knapsack DP tests passed successfully.")

File: demos/javascript/dsa/dp/knapsack.js
function knapsack01(W, weights, values) {
  const n = weights.length;
  const dp = new Int32Array(W + 1);

  for (let i = 0; i < n; i++) {
    for (let w = W; w >= weights[i]; w--) {
      dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
    }
  }
  return dp[W];
}

console.log("=== JavaScript 0/1 Knapsack DP ===");
const weights = [2, 3, 4, 5];
const values = [3, 4, 5, 6];
const maxVal = knapsack01(5, weights, values);
console.assert(maxVal === 7, "Knapsack assertion failed");
console.log("Max Knapsack Value for W=5:", maxVal);
console.log("JavaScript Knapsack DP tests passed successfully.");
